remote write: reloading config with modified queue parameters results in duplicate metric registration

**What did you do?**

Changed remote write `queue_config` fields like `max_shards`, `capacity`, `max_samples_per_send`.
SIGHUP'd the server.

**What did you expect to see?**

New config to be applied.

**What did you see instead? Under which circumstances?**

A panic.

After the following restart, the server came back OK. It's just that with a runtime config reload, the metrics of the new remote write queues collide with the old metrics.

Somehow, the unregister call added in #7188 isn't called in time or doesn't work as expected.

**Environment**

* Prometheus version:

	2.19.1

* Logs:
```
2020-06-24 15:02:25	
	/app/vendor/github.com/oklog/run/group.go:37 +0xbb
2020-06-24 15:02:25	
created by github.com/oklog/run.(*Group).Run
2020-06-24 15:02:25	
	/app/vendor/github.com/oklog/run/group.go:38 +0x27
2020-06-24 15:02:25	
github.com/oklog/run.(*Group).Run.func1(0xc000b001e0, 0xc000b00120, 0xc000912000)
2020-06-24 15:02:25	
	/app/cmd/prometheus/main.go:607 +0x1ec
2020-06-24 15:02:25	
main.main.func14(0x0, 0x0)
2020-06-24 15:02:25	
	/app/cmd/prometheus/main.go:817 +0x262
2020-06-24 15:02:25	
main.reloadConfig(0x7fffeda61ddc, 0x1e, 0x2f3b7e0, 0xc000c72cc0, 0xc000620340, 0x8, 0x8, 0x0, 0x0)
2020-06-24 15:02:25	
	/app/storage/remote/storage.go:77 +0xa6
2020-06-24 15:02:25	
github.com/prometheus/prometheus/storage/remote.(*Storage).ApplyConfig(0xc000188100, 0xc3d026fc00, 0x0, 0x0)
2020-06-24 15:02:25	
	/app/storage/remote/write.go:141 +0x3eb
2020-06-24 15:02:25	
github.com/prometheus/prometheus/storage/remote.(*WriteStorage).ApplyConfig(0xc000bcaa00, 0xc3d026fc00, 0x0, 0x0)
2020-06-24 15:02:25	
	/app/storage/remote/queue_manager.go:181 +0x10f2
2020-06-24 15:02:25	
github.com/prometheus/prometheus/storage/remote.newQueueManagerMetrics(0x2f76760, 0xc00019e730, 0xc755015ec0, 0x13, 0xc9885a4100, 0x40, 0xc771a42301)
2020-06-24 15:02:25	
	/app/vendor/github.com/prometheus/client_golang/prometheus/registry.go:400 +0xad
2020-06-24 15:02:25	
github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc00019e730, 0xc0921c6000, 0xe, 0xe)
2020-06-24 15:02:25	
goroutine 75 [running]:
2020-06-24 15:02:25	

2020-06-24 15:02:25	
panic: duplicate metrics collector registration attempted
2020-06-24 15:02:25	
level=info ts=2020-06-24T13:02:25.313Z caller=main.go:799 msg="Loading configuration file" filename=/etc/prometheus/prometheus.yml

```

--- a/storage/remote/queue_manager.go
+++ b/storage/remote/queue_manager.go
@@ -177,8 +177,12 @@ func newQueueManagerMetrics(r prometheus.Registerer, rn, e string) *queueManager
 		ConstLabels: constLabels,
 	})
 
-	if r != nil {
-		r.MustRegister(
+	return m
+}
+
+func (m *queueManagerMetrics) register() {
+	if m.reg != nil {
+		m.reg.MustRegister(
 			m.succeededSamplesTotal,
 			m.failedSamplesTotal,
 			m.retriedSamplesTotal,
@@ -195,7 +199,6 @@ func newQueueManagerMetrics(r prometheus.Registerer, rn, e string) *queueManager
 			m.bytesSent,
 		)
 	}
-	return m
 }
 
 func (m *queueManagerMetrics) unregister() {
@@ -358,7 +361,8 @@ outer:
 // Start the queue manager sending samples to the remote storage.
 // Does not block.
 func (t *QueueManager) Start() {
-	// Initialise some metrics.
+	// Register and initialise some metrics.
+	t.metrics.register()
 	t.metrics.shardCapacity.Set(float64(t.cfg.Capacity))
 	t.metrics.pendingSamples.Set(0)
 	t.metrics.maxNumShards.Set(float64(t.cfg.MaxShards))

--- a/storage/remote/write_test.go
+++ b/storage/remote/write_test.go
@@ -20,6 +20,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
 	common_config "github.com/prometheus/common/config"
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/config"
@@ -154,6 +155,48 @@ func TestRestartOnNameChange(t *testing.T) {
 	testutil.Ok(t, err)
 }
 
+func TestUpdateWithRegisterer(t *testing.T) {
+	dir, err := ioutil.TempDir("", "TestRestartWithRegisterer")
+	testutil.Ok(t, err)
+	defer os.RemoveAll(dir)
+
+	s := NewWriteStorage(nil, prometheus.NewRegistry(), dir, time.Millisecond)
+	c1 := &config.RemoteWriteConfig{
+		Name: "named",
+		URL: &common_config.URL{
+			URL: &url.URL{
+				Scheme: "http",
+				Host:   "localhost",
+			},
+		},
+		QueueConfig: config.DefaultQueueConfig,
+	}
+	c2 := &config.RemoteWriteConfig{
+		URL: &common_config.URL{
+			URL: &url.URL{
+				Scheme: "http",
+				Host:   "localhost",
+			},
+		},
+		QueueConfig: config.DefaultQueueConfig,
+	}
+	conf := &config.Config{
+		GlobalConfig:       config.DefaultGlobalConfig,
+		RemoteWriteConfigs: []*config.RemoteWriteConfig{c1, c2},
+	}
+	testutil.Ok(t, s.ApplyConfig(conf))
+
+	c1.QueueConfig.MaxShards = 10
+	c2.QueueConfig.MaxShards = 10
+	testutil.Ok(t, s.ApplyConfig(conf))
+	for _, queue := range s.queues {
+		testutil.Equals(t, 10, queue.cfg.MaxShards)
+	}
+
+	err = s.Close()
+	testutil.Ok(t, err)
+}
+
 func TestWriteStorageLifecycle(t *testing.T) {
 	dir, err := ioutil.TempDir("", "TestWriteStorageLifecycle")
 	testutil.Ok(t, err)
@@ -178,7 +221,7 @@ func TestUpdateExternalLabels(t *testing.T) {
 	testutil.Ok(t, err)
 	defer os.RemoveAll(dir)
 
-	s := NewWriteStorage(nil, nil, dir, time.Second)
+	s := NewWriteStorage(nil, prometheus.NewRegistry(), dir, time.Second)
 
 	externalLabels := labels.FromStrings("external", "true")
 	conf := &config.Config{

